core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -249,6 +253,12 @@ private FileSystemContext(AlluxioConfiguration conf, @Nullable BlockWorker block
     mBlockWorker = blockWorker;
     mWorkerRefreshPolicy =
         new TimeoutRefresh(conf.getMs(PropertyKey.USER_WORKER_LIST_REFRESH_INTERVAL));
+    mReadBlockLocationPolicy =
+        BlockLocationPolicy.Factory.create(
+            conf.getClass(PropertyKey.USER_UFS_BLOCK_READ_LOCATION_POLICY), conf);
+    mWriteBlockLocationPolicy =
+        BlockLocationPolicy.Factory.create(
+            conf.getClass(PropertyKey.USER_BLOCK_WRITE_LOCATION_POLICY), conf);
```

Comment:
If you decide the policy here, the path config are not really resolved yet
https://github.com/Alluxio/alluxio/blob/5828d56a824748e7c91076842fad75efb42f92f9/core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java#L171

This is my biggest concern now and I think other parts of this PR look good.

Reply:
`RoundRobinPolicy` caches the worker infos the first time its `getWorker` method gets called:

https://github.com/Alluxio/alluxio/blob/da543a5b604cae9703d55dad4102996ab685c3b1/core/client/fs/src/main/java/alluxio/client/block/policy/RoundRobinPolicy.java#L75-L79

When the policy's lifetime is tied to a specific file stream, it's totally fine. But if it's tied to the lifetime of the client, the information can get stale and is likely to cause unavailable exceptions.

Reply:
How about round-robin selection based on worker address hashcodes?

First add a new field:
```java
  // initialize with a random integer to start with a random worker
  private int mPreviousHashCode = new Random().nextInt();
```

Then implement the round-robin algorithm as:

```java

    // Round-robin selection based on worker address hashcodes
    // The assumption is that WorkerNetAddress has a well-implemented `hashCode()`
    // that ensures different workers have distinct hashcodes.

    // find the worker whose address object has a hashcode that is greater than
    // the previous selection but the smallest among all.
    Optional<WorkerNetAddress> candidate = Streams.stream(options.getBlockWorkerInfos())
        .map(BlockWorkerInfo::getNetAddress)
        .filter(workerNetAddress -> workerNetAddress.hashCode() > mPreviousHashCode)
        .min(Comparator.comparingInt(WorkerNetAddress::hashCode));
    if (!candidate.isPresent()) {
      // reset the hashcode and pick the smallest one
      Optional<WorkerNetAddress> smallest = Streams.stream(options.getBlockWorkerInfos())
          .map(BlockWorkerInfo::getNetAddress)
          .min(Comparator.comparingInt(WorkerNetAddress::hashCode));
      mPreviousHashCode = smallest.map(WorkerNetAddress::hashCode).orElse(Integer.MIN_VALUE);
      return smallest;
    } else {
      mPreviousHashCode = candidate.get().hashCode();
      return candidate;
    }
```

Reply:
> `RoundRobinPolicy` caches the worker infos the first time its `getWorker` method gets called:
> 
> https://github.com/Alluxio/alluxio/blob/da543a5b604cae9703d55dad4102996ab685c3b1/core/client/fs/src/main/java/alluxio/client/block/policy/RoundRobinPolicy.java#L75-L79
> 
> When the policy's lifetime is tied to a specific file stream, it's totally fine. But if it's tied to the lifetime of the client, the information can get stale and is likely to cause unavailable exceptions.

Is this getting addressed in a separate PR?

Reply:
This can be addressed in a separate PR later

Reply:
#16228